Hyperdrive 0.2.6 (#30)
p2pcommons/specs: module.md

## Initialization

One module is defined as a [Hyperdrive](https://github.com/mafintosh/hyperdrive/) (a folder on a machine)
containing [module metadata](#module-metadata) in accordance with this
specification.

> A Hyperdrive is considered initialized once it has been validated

Reviewer: ```suggestion
A Hyperdrive is initialized when it has emitted the "ready" event
and thus has a public key (64 character hash). We utilize Hyperdrive `v10`.
```

and has a public key (64 character hash). We utilize Hyperdrive `v10`
cores.

In general, Hyperdrive keys in this specification MAY be suffixed by

> `+{vers}` where `vers` is an integer denoting the Hyperdrive version. 

Reviewer: A hyperdrive key itself won't be suffixed, it's rather `key+version` instead, for which we need a name. I think we called it `versioned key`. So it could be a key or a versioned key I think?

Author: How about:

A versioned Hyperdrive key MUST be suffixed with `+{vers}` where `vers`
is an integer denoting the Hyperdrive version. A non-versioned Hyperdrive
key MUST NOT be suffixed with a version.

Reviewer: sounds good!

## Initialization

One module is defined as a [Hyperdrive](https://github.com/mafintosh/hyperdrive/) (a folder on a machine)
containing [module metadata](#module-metadata) in accordance with this
specification.

A Hyperdrive is initialized when it has emitted the "ready" event
and thus has a public key (64 character hash). We utilize Hyperdrive `v10`.

> and has a public key (64 character hash). We utilize Hyperdrive `v10`
> cores.

Reviewer: ```suggestion
```

| `p2pcommons.main`     | string           | [`^((?!\/) OR (\.\/))(?!~ OR \.).*(?<!\/)$`](https://regex101.com/r/MZXJnK/1)                       |
| `p2pcommons.avatar`     | string           | [`^((?!\/) OR (\.\/))(?!~ OR \.).*(?<!\/)$`](https://regex101.com/r/MZXJnK/1)                       |
| `p2pcommons.authors`  | array of strings | [`^(\w{64})$`](https://regex101.com/r/GQOim5/2)                                          |
| `p2pcommons.parents`  | array of strings | [`^(\w{64})(\+\d+)$`](https://regex101.com/r/GQOim5/3)                                   |
| `p2pcommons.follows`  | array of strings | [`^(\w{64})(\+\d+)?$`](https://regex101.com/r/GQOim5/4)                                  |
| `p2pcommons.contents` | array of strings | [`^(\w{64})(\+\d+)?$`](https://regex101.com/r/GQOim5/4)                                  |

`title` and `description` MUST be strings. `title` MUST contain a
string between 1 and 300 characters long and MUST NOT consist only
of whitespace characters. `description` MAY be empty (`''`; see also
[Registration](#registration)).

> `url` MUST be a string containing the non-versioned Hyperdrive key

Reviewer: This isn't an `url` any more, it's a `key` now, this field needs to be renamed.

`links.spec` MUST contain an object in an array with key/value pair
`href` referring to the active module specification URL (e.g.,

> `https://p2pcommons.com/specs/module/0.2.6`).

Reviewer: instead of updating the version multiple times in the document, what about always referring to the page header?

Author: It's a few seconds extra work each time and is OK to forget I'd say. How would you suggest referring to the page header?

Reviewer: ok sorry the referring part is wrong here, what about just letting these go out of date? It's not just about updating them, it also adds noise to the pull request review.

> `p2pcommons.main` MUST be a string containing one relative path but MUST
> NOT refer to a relative home or relative parent directory. The

Reviewer 2: ```suggestion
`p2pcommons.main` MUST be a string containing one relative path but MUST
NOT refer to a relative home or relative parent directory and MUST NOT refer to a dotfile (e.g., `./.example.json`). The
```

Reviewer 2: This is outside of the original PR scope, thought I'd simply add it in now given it was on the to-do list 👍

Author: Wouldn't changing `url` break with https://github.com/datprotocol/dat.json and therefore currently break interoperability?

I would also suggest leaving the `.json` off the dotfile example, as (afaik) dotfiles usually don't have an extension.

Reviewer: > I would also suggest leaving the `.json` off the dotfile example, as (afaik) dotfiles usually don't have an extension.

In my experience this is not true

Reviewer: > Wouldn't changing `url` break with https://github.com/datprotocol/dat.json and therefore currently break interoperability?

Great point! They use `url` and have a protocol in there too.

Not sure what to do here..if we go back to urls we have to add back the protocol again, and we don't know it yet.

Author: The Dat spec says that it will "usually" be the `dat://` protocol prefix. We previously had the protocol optional in the spec, so I don't see any risk in decreased interoperability if we now leave out the protocol ourselves but leave it as `url`.

Reviewer: Technically an url MUST contain a prefix, see https://en.wikipedia.org/wiki/URL#Syntax.

Reviewer: - a) means more work and potentially bad side effects
- b) is less correct but less work and keeps interoperability
- c) hurts (:P)

This would be easier if we didn't aim for spec stability soon, as we can't tell how the dat/hyper system will evolve. @chartgerink what do you think?

Author: Of course we can choose to have the `url` contain `dat://` but have all our custom fields be protocol-free. That would reduce effort when protocol changes are made.

Reviewer 3: I'm in for waiting a bit to see how the ecosystem evolves like julian says. `hyper://` apps are quite new. Personally I'd like to see this as a key and that's it because it is what it is. And we are not using in any other form yet.

Reviewer 2: I propose to keep it as `url` with the `dat://` for now and revisit in four weeks? @jameslibscie this would agree with @dpaez and @juliangruber to a certain degree, but would require undoing the changes you made for now.

Four weeks gives us some time to think about this and see how things evolve. Stability is key when we bump to `v1.0` of the specs, and it seems like it is important to understand how the landscape develops. 

I agree with @juliangruber that we are aiming for stability quite soon, but it is necessary given the timeframe we are working in. Otherwise we are kicking the can down the road and `v1.0` of the SDK-JS is affected, which affects `v1.0` of Hypergraph. Maybe it is worth reaching out to Paul/Mathias/Andrew and asking their take on the landscape developments?

Author: I agree with this, especially if it's just the changes to `url`. I think we can safely remove the protocol from all of our custom fields.

Reviewer: asked on twitter: https://twitter.com/juliangruber/status/1255109070429868033

Reviewer: `hyper://` confirmed: https://twitter.com/mafintosh/status/1255121694127280128 :) so we can switch back to the `url`

Author: I've updated `dat://` to `hyper://` for the `url`. It's now also no longer optional to include the prefix.

Before, the protocol was an optional part of the key, but now it isn't part of the key at all, so I've updated the description of `url` to reflect this.

  "url": "00a4f2f18bb6cb4e9ba7c2c047c8560d34047457500e415d535de0526c6b4f23",
  "links": {
     "license": [{"href": "https://creativecommons.org/publicdomain/zero/1.0/legalcode"}],

>      "spec": [{"href": "https://p2pcommons/specs/module/0.2.6"}]

Reviewer: I think this can stay as `0.2.2`, so we don't have the maintenance overhead (also below)
